src/libcore/slice/mod.rs

```diff
@@ -581,7 +607,7 @@ impl<T> [T] {
     pub fn iter(&self) -> Iter<T> {
         unsafe {
             let p = if mem::size_of::<T>() == 0 {
-                1 as *const _
+                NonNull::dangling().as_ptr() as *const _
```

Comment:
This makes me think that [`Iter::ptr`](https://github.com/rust-lang/rust/blob/b3e7d70ce7730a929fd3d5094e8ad2eb71c3a399/src/libcore/slice/mod.rs#L2600) should be a `NonNull<T>` instead of a `*const T`...

Reply:
It could be. Same for `end`. (Though I have plans for #42789 that would need tweaking of `end` were to become a `NonNull`.)

Do you want me to change that in this PR?

Reply:
I have no strong feelings about it.  If it's easy and it's mostly lines that were changing anyway, might as well; if it's annoying or makes the PR a bunch longer, then probably not worth doing right now.  (`end` not being `NonNull` is also totally fine by me.)

Reply:
It's annoying because tons of code here does raw pointer operations, so there'd be a lot of `to_ptr` and `new_unchecked` and so on